File: model_1/icpc-jakarta-2018/future_1_solution.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

int main() {
    int N;
    cin >> N;
    vector<string> S(N);
    for (int i = 0; i < N; i++) {
        cin >> S[i];
    }

    vector<vector<string>> A_list(N);
    for (int i = 0; i < N; i++) {
        string s = S[i];
        set<string> seen;
        int n = s.length();
        for (int mask = 1; mask < (1 << n); mask++) {
            string t = "";
            for (int j = 0; j < n; j++) {
                if (mask & (1 << j)) {
                    t += s[j];
                }
            }
            seen.insert(t);
        }
        vector<string> A(seen.begin(), seen.end());
        sort(A.begin(), A.end());
        A_list[i] = A;
    }

    vector<int> dp_prev;
    for (string t : A_list[0]) {
        dp_prev.push_back(t.length());
    }

    for (int i = 1; i < N; i++) {
        vector<string> prev_list = A_list[i-1];
        vector<int> prefix_max(dp_prev.size(), 0);
        if (dp_prev.size() > 0) {
            prefix_max[0] = dp_prev[0];
            for (int j = 1; j < dp_prev.size(); j++) {
                prefix_max[j] = max(prefix_max[j-1], dp_prev[j]);
            }
        }

        vector<string> current_list = A_list[i];
        vector<int> dp_current(current_list.size(), -1000000000);

        for (int j = 0; j < current_list.size(); j++) {
            string t = current_list[j];
            auto it = lower_bound(prev_list.begin(), prev_list.end(), t);
            int pos = it - prev_list.begin();
            if (pos > 0) {
                dp_current[j] = prefix_max[pos-1] + t.length();
            }
        }
        dp_prev = dp_current;
    }

    int ans = -1;
    if (!dp_prev.empty()) {
        ans = *max_element(dp_prev.begin(), dp_prev.end());
        if (ans < 0) {
            ans = -1;
        }
    } else {
        ans = -1;
    }

    cout << ans << endl;

    return 0;
}

File: model_1/icpc-jakarta-2018/future_1_solution.py
import sys

def main():
    data = sys.stdin.read().splitlines()
    if not data:
        return
    N = int(data[0].strip())
    S = []
    for i in range(1, 1 + N):
        S.append(data[i].strip())
    
    A_list = []
    for s in S:
        n = len(s)
        seen = set()
        for mask in range(1, 1 << n):
            substr = []
            for j in range(n):
                if mask & (1 << j):
                    substr.append(s[j])
            t = ''.join(substr)
            seen.add(t)
        sorted_seen = sorted(seen)
        A_list.append(sorted_seen)
    
    if N == 0:
        print(0)
        return
        
    dp_prev = [len(subseq) for subseq in A_list[0]]
    
    for i in range(1, N):
        prev_A = A_list[i-1]
        current_A = A_list[i]
        j = 0
        best_so_far = -10**9
        current_dp = [-10**9] * len(current_A)
        
        for idx in range(len(current_A)):
            s = current_A[idx]
            while j < len(prev_A) and prev_A[j] < s:
                if dp_prev[j] > best_so_far:
                    best_so_far = dp_prev[j]
                j += 1
            
            if best_so_far != -10**9:
                current_dp[idx] = best_so_far + len(s)
        
        dp_prev = current_dp
        
    ans = max(dp_prev) if dp_prev else -10**9
    if ans < 0:
        print(-1)
    else:
        print(ans)

if __name__ == "__main__":
    main()